File: c++/number-of-islands.cc
class Solution {
public:
    void exploreIsland(vector<vector<char>>& grid, int m, int n, int x, int y) {
        stack<pair<int, int>> s;
        s.push({x, y});
        while (!s.empty()) {
            auto node = s.top();
            s.pop();

            grid[node.second][node.first] = '0';
            if (node.first + 1 < m && grid[node.second][node.first + 1] == '1') {
                s.push({node.first + 1, node.second});
            }
            if (node.first - 1 > -1 && grid[node.second][node.first - 1] == '1') {
                s.push({node.first - 1, node.second});
            }
            if (node.second + 1 < n && grid[node.second + 1][node.first] == '1') {
                s.push({node.first, node.second + 1});
            }
            if (node.second - 1 > -1 && grid[node.second - 1][node.first] == '1') {
                s.push({node.first, node.second - 1});
            }
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        int m = grid[0].size();
        int n = grid.size();

        int islandCount = 0;

        for (int i = 0; i < n; i ++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '0') continue;
                islandCount++;
                exploreIsland(grid, m, n, j, i);
            }
        }
        return islandCount;
    }
};

File: python/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def exploreIsland(x: int, y: int) -> None:
            if y < 0 or x < 0 or y >= len(grid) or x >= len(grid[0]) or grid[y][x] == '0':
                return
            grid[y][x] = '0'
            exploreIsland(x + 1, y)
            exploreIsland(x - 1, y)
            exploreIsland(x, y + 1)
            exploreIsland(x, y - 1)
        
        numIslands = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    numIslands += 1
                    exploreIsland(j, i)
        
        return numIslands
